core/modules/contact/contact.admin.inc

```diff
@@ -53,9 +62,95 @@ function contact_settings_form($form, &$form_state) {
     '#default_value' => $config->get('contact_threshold_window'),
   );
 
+  $form['#submit'][] = 'contact_settings_form_submit';
   return system_settings_form($form);
 }
 
+/**
+ * Contact settings form submit handler.
+ *
+ * @see contact_settings_form()
+ */
+function contact_settings_form_submit($form, &$form_state) {
+  // Update user permissions for each role.
+  foreach ($form_state['values']['permissions']['roles'] as $role_name => $role) {
+    user_role_change_permissions($role->name, $form_state['values']['permissions']['checkboxes'][$role->name]);
+  }
+
+  // Remove permissions from default #config store (contact.settings).
+  unset($form_state['values']['permissions']);
+}
+
+/**
+ * Builds a matrix of contact permissions.
+ *
+ * @see node_type_form_permissions()
+ */
+function contact_form_permissions() {
+  // Retrieve role names for columns.
+  $roles = user_roles(FALSE, NULL, TRUE);
+  $admin_role = config_get('system.core', 'user_admin_role');
+
+  // Note that in contact_theme(), the 'contact_form_permissions' theme callback
+  // reuses theme_user_admin_permissions(), to make it identical to the user
+  // module permissions page.
+  $form['#theme'] = array('contact_form_permissions');
+
+  // Wrap the table in a form element to make margins match other form elements.
+  $form['#theme_wrappers'] = array('form_element');
+
+  $form['roles'] = array(
+    '#type' => 'value',
+    '#value' => $roles,
+  );
+
+  // Retrieve contact related permissions.
+  $permissions = contact_permission();
+
+  // Render role/permission overview:
+  $options = array();
+  $status = array();
+  foreach ($permissions as $perm => $perm_item) {
```

Comment:
I would use full names for variables. Makes them less "cryptic" and avoids ambiguities when these variables are being used later in the code (perm = "permission" or "permanent").